### PR TITLE
Address indexing OOB in darcy datapipe

### LINKED ISSUE #1958: 🐛[BUG]: Darcy2D multigrid kernels read out of bounds and can return huge/NaN truth
### Version

PhysicsNeMo `2.3.0a0`, commit [`4fbfcfd62bf050b48ceec6b438da409b9f4644b3`](https://github.com/NVIDIA/physicsnemo/commit/4fbfcfd62bf050b48ceec6b438da409b9f4644b3) (current `main` when retested on 2026-09-02).

### On which installation method(s) does this occur?

Source (`uv`, editable checkout).

### Describe the issue

`Darcy2D` can return silently corrupted pressure fields when `nr_multigrids >= 3`. Normalized `darcy` values are normally about `-1.6` to `3.8`, but affected runs return finite values around `1e24` to `1e31`, or NaNs. The finite outliers overflow float32 MSE and make training loss `inf`.

The problem is an out-of-bounds read in the multigrid Jacobi path, rather than FNO training or a particular random permeability:

- For `resolution=64`, the buffers have shape `(batch, 65, 65)`.
- [`darcy_mgrid_jacobi_iterative_batched_2d`](https://github.com/NVIDIA/physicsnemo/blob/4fbfcfd62bf050b48ceec6b438da409b9f4644b3/physicsnemo/datapipes/benchmarks/kernels/finite_difference.py#L69-L85) maps the last coarse-grid point to index 63 and then reads the positive neighbor at 71 for factor 8, or 67 for factor 4.
- [`index_zero_edges_batched_2d`](https://github.com/NVIDIA/physicsnemo/blob/4fbfcfd62bf050b48ceec6b438da409b9f4644b3/physicsnemo/datapipes/benchmarks/kernels/indexing.py#L80-L89) handles only indices exactly equal to `-1`, `lx`, or `ly`. Indices 67 and 71 therefore fall through to `array[b, x, y]` even though the largest valid index is 64.
- Warp debug mode consistently aborts at the resulting array bounds check. In release mode, the result depends on adjacent allocation contents, explaining the intermittent behavior.

There also appears to be a related factor-4/8 interpolation issue in [`bilinear_upsample_batched_2d`](https://github.com/NVIDIA/physicsnemo/blob/4fbfcfd62bf050b48ceec6b438da409b9f4644b3/physicsnemo/datapipes/benchmarks/kernels/utils.py#L53-L62): `x_1 = x + (x + 1) % factor` (and the analogous `y_1`) does not select the next coarse node for factors greater than 2. The kernel reads and writes the same array in one launch as well.

Expected behavior: every returned Darcy field is finite and within a physically/numerically plausible range, or the datapipe reports solver failure instead of returning an unconverged/corrupted field.

Actual behavior from the normal FNO training example on current `main` (five fresh sequential processes): one process failed on batch 0 while four appeared normal:

```text
pred min/max: 0.020137101411819458 0.022842921316623688
truth min/max: 1.9012562899356852e+26 8.182267788361775e+28
loss is non-finite
```

### Minimum reproducible example

The following canary makes the release-mode failure deterministic. It keeps the original datapipe and solver kernels, but aliases the logical `65 x 65` Darcy buffers into padded allocations whose halo is filled with `1e20`. A bounds-correct kernel cannot observe the halo.

```python
import numpy as np
import torch
import warp as wp

from physicsnemo.datapipes.benchmarks.darcy import Darcy2D

LOGICAL_SIZE = 65
PADDED_SIZE = 73  # Includes the factor-8 positive neighbor at index 71.
CANARY = 1.0e20


def logical_view(parent: wp.array) -> wp.array:
    return wp.array(
        ptr=parent.ptr,
        dtype=float,
        shape=(1, LOGICAL_SIZE, LOGICAL_SIZE),
        strides=parent.strides,
        capacity=parent.capacity,
        device="cpu",
    )


np.random.seed(42)
torch.manual_seed(42)

datapipe = Darcy2D(
    resolution=64,
    batch_size=1,
    nr_permeability_freq=5,
    max_iterations=1000,
    convergence_threshold=1.0e-4,
    iterations_per_convergence_check=50,
    nr_multigrids=4,
    normaliser={
        "permeability": (1.25, 0.75),
        "darcy": (4.52e-2, 2.79e-2),
    },
    device="cpu",
)

darcy0_parent = wp.full(
    (1, PADDED_SIZE, PADDED_SIZE), CANARY, dtype=float, device="cpu"
)
darcy1_parent = wp.full(
    (1, PADDED_SIZE, PADDED_SIZE), CANARY, dtype=float, device="cpu"
)
datapipe.darcy0 = logical_view(darcy0_parent)
datapipe.darcy1 = logical_view(darcy1_parent)

batch = next(iter(datapipe))
truth = batch["darcy"]
print(
    f"permeability=[{float(batch['permeability'].min())}, "
    f"{float(batch['permeability'].max())}]"
)
print(
    f"truth=[{float(truth.min())}, {float(truth.max())}] "
    f"max_abs={float(truth.abs().max())} "
    f"finite={bool(torch.isfinite(truth).all())}"
)
```

Run with:

```shell
uv run --no-sync python darcy2d_oob_canary.py
```

To reproduce the bounds assertion directly, set `wp.config.mode = "debug"` before constructing `Darcy2D` and request one batch with `nr_multigrids=4`.

### Relevant log output

Deterministic release-mode canary output:

```text
permeability=[-1.0, 1.0]
truth=[2.1828058885390336e+16, 9.659499921081893e+18]
max_abs=9.659499921081893e+18 finite=True
```

Warp debug-mode output:

```text
Assertion failed: 'k >= -arr.shape[2] && k < arr.shape[2]'
At '.../site-packages/warp/native/array.h:478'
```

Exact indices for this configuration:

| reduction factor | last solved index | positive neighbor | valid upper bound |
|---:|---:|---:|---:|
| 8 | 63 | 71 | 64 |
| 4 | 63 | 67 | 64 |
| 2 | 63 | 65 | boundary sentinel handled by helper |
| 1 | 64 | 65 | boundary sentinel handled by helper |

As a temporary workaround, `nr_multigrids=2` avoids the factor-4/8 accesses. Twenty batches on current `main`, using `max_iterations=5000` and `convergence_threshold=1e-6`, were all finite with normalized `max_abs <= 2.643`.

Suggested fixes/tests:

1. Have zero-boundary indexing reject the complete out-of-range domain (`x < 0 or x >= lx or y < 0 or y >= ly`).
2. Correct factor-4/8 interpolation coordinates and avoid in-place source/destination aliasing during upsampling.
3. Report failure when a multigrid level exhausts `max_iterations`, and reject non-finite residuals/solutions.
4. Add Warp CPU debug-mode tests for `nr_multigrids=3/4` and interpolation reference tests for factors 2/4/8.

### Environment details

```text
Environment: bare-metal macOS, Apple Silicon arm64, CPU backend
macOS: 15.7.9
Python: 3.13.1
PhysicsNeMo: 2.3.0a0, source commit 4fbfcfd62bf050b48ceec6b438da409b9f4644b3
Warp: 1.14.0
PyTorch: 2.12.0
NumPy: 2.4.6
uv: 0.11.29
CUDA: unavailable
MPS: not used (Warp runs on CPU)
```

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

Fix indexing issues in the Darcy datapipes, which would let kernels read past the ends of arrays in some cases.  also uses out-of-place upsampling now, instead of inplace.

Closes #1958 .

## Checklist

- [X] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.
- [X] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [X] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

**Review by peterdsharpe — APPROVED** [review]
Looks good, yep

**Comment by coreyjadams** [comment]
/ok to test [d7c6d17](https://github.com/NVIDIA/physicsnemo/pull/1960/commits/d7c6d17c54fffc072fd4e3e9d210947529592ddf)

**Review by negin513 — APPROVED** [review]
